pyam/core.py

```diff
@@ -733,118 +727,140 @@ def check_aggregate(self, variable, components=None, units=None,
         kwargs: passed to `np.isclose()`
         """
         # compute aggregate from components, return None if no components
-        df_components = self.aggregate(variable, components, units)
+        df_components = self.aggregate(variable, components, unit)
         if df_components is None:
             return
 
         # filter and groupby data, use `pd.Series.align` for matching index
+        rows = self._apply_filters(variable=variable, unit=unit)
         df_variable, df_components = (
-            _aggregate_by_variables(self.data, variable, units)
-            .align(df_components)
+            _aggregate(self.data[rows], 'variable').align(df_components)
         )
 
         # use `np.isclose` for checking match
         diff = df_variable[~np.isclose(df_variable, multiplier * df_components,
                                        **kwargs)]
 
         if len(diff):
-            msg = '{} - {} of {} data points are not aggregates of components'
+            msg = '`{}` - {} of {} rows are not aggregates of components'
             logger().info(msg.format(variable, len(diff), len(df_variable)))
 
             if exclude_on_fail:
                 self._exclude_on_fail(diff.index.droplevel([2, 3, 4]))
 
-            diff = pd.concat([diff], keys=[variable], names=['variable'])
+            return IamDataFrame(diff, variable=variable).timeseries()
+
+    def aggregate_region(self, variable, region='World', subregions=None,
```

Comment:
A few thoughts here:

1) Instead of a region/subregion argument, there should be something like `mapping` which is a dict, dataframe, etc. with key-value pairs of existing-region to aggregate region
2) `unit` here seems to be a filtering argument, I think that should not be here but used explicitly with `filter` before hand
3) my gut feeling is that `component` here is extraneous (a user should instead do df.aggregate_region().aggregate_variable()`, but I think based on our discussion this is unavoidable - let me know what you think

Reply:
Re 1: seems to me a question of personal preference whether to pass a dict or explicit args. Question would be how to ensure that the dict-way is meaningful (allow overlapping region aggregation?, order of aggregation if you have multiple levels of region hierarchy?).

Reply:
Re 3: note that this function adds variable components that only exist at the aggregate-region Level to the new timeseries, so this has to be part of this function.

Reply:
> 1. Instead of a region/subregion argument, there should be something like `mapping` which is a dict, dataframe, etc. with key-value pairs of existing-region to aggregate region

This would allow you to do multiple region aggregations at once right? If so, perhaps that could be in a helper method (i.e. this method would stay, and would always just do one region at a time, and we might have another, `aggregate_regions` which would allow you to define a mapping in order to do many regions, all to different 'targets', at once).

> 2\. `unit` here seems to be a filtering argument, I think that should not be here but used explicitly with `filter` before hand

+1

> 3\. my gut feeling is that `component` here is extraneous (a user should instead do df.aggregate_region().aggregate_variable()`, but I think based on our discussion this is unavoidable - let me know what you think

Thinking about it more, I'm not sure that this argument is actually wise. The only case I can think of where you'd want it is if you wanted to introduce an inconsistency into your dataset (e.g. you wanted 'CO2|Air' not to be included in 'CO2' at the world level). I'm not sure if we should allow this. 

If you just drop this keyword, then you ensure that the data stays internally consistent and if users want to do something else, that's up to them.

Reply:
Re 1: we could refactor `map_regions()` to take a mapping, do some consistency checks (e.g., regions exist?), and then call `aggregate_region()` iteratively or (in the future `downscale_region()`), and return the concatenated `IamDataFrame` of the new region mapping (or append to `self`). That would make for multiple shorter functions, rather than one beast of a function...

Reply:
Re @znicholls and 3: unfortunately, we cannot force IAM users to be consistent in their variable tree hierarchy, so we will have variables like `Primary Energy|Coal` and `Primary Energy|Fossil` in one `IamDataFrame` until at least IPCC AR8 (*). So the function needs to have a somewhat decent way to handle such cases - and requiring the user to provide a custom list of `components` seems to me the easiest way.

(*) Yes, one could start `Primary Energy|Fossil|Coal`. Not gonna happen...

Reply:
> unfortunately, we cannot force IAM users to be consistent in their variable tree hierarchy

ah yes I forgot about this. Keeping `components` sounds good then

Reply:
On 1, I would think it would be pretty straightforward to use [`pd.Series.map`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.map.html) to change region names as desired, then `groupby().sum()`. My suggestion here is that we do our best to let pandas take care of error messaging, etc. for us. I really think we should not try to reinvent the wheel, and pass as much as we can directly to pandas methods.

Reply:
what was the resolution here? Use a dict map as a refactoring in future?

Reply:
For now, we will refactor `map_regions()` to take a mapping ad call `aggregate_region()` or `downscale_region()´ iteratively, see #225.

Later on, extending `aggregate_region()` to take a `mapping` instead of the string args `region` and `subregions` may be practical for speedup.